src/index.ts

```diff
@@ -153,35 +154,33 @@ async function runTests(
 
     if (onlyTestTsNext || tsLocal) {
         const tsVersion = tsLocal ? "local" : TypeScriptVersion.latest;
-        if (typesVersions.length === 0) {
-            await testTypesVersion(dirPath, tsVersion, tsVersion, isOlderVersion, dt, indexText, expectOnly, tsLocal);
-        } else {
-            const latestTypesVersion = last(typesVersions);
-            const versionPath = joinPaths(dirPath, `ts${latestTypesVersion}`);
-            const versionIndexText = await readFile(joinPaths(versionPath, "index.d.ts"), "utf-8");
-            await testTypesVersion(
-                versionPath, tsVersion, tsVersion,
-                isOlderVersion, dt, versionIndexText, expectOnly, tsLocal, /*inTypesVersionDirectory*/ true);
-        }
+        await testTypesVersion(dirPath, tsVersion, tsVersion, isOlderVersion, dt, indexText, expectOnly, tsLocal, /*isLatest*/ true);
     } else {
-        await testTypesVersion(dirPath, undefined, getTsVersion(0), isOlderVersion, dt, indexText, expectOnly, undefined);
-        for (let i = 0; i < typesVersions.length; i++) {
-            const version = typesVersions[i];
-            const versionPath = joinPaths(dirPath, `ts${version}`);
-            const versionIndexText = await readFile(joinPaths(versionPath, "index.d.ts"), "utf-8");
-            await testTypesVersion(
-                versionPath, version, getTsVersion(i + 1), isOlderVersion, dt, versionIndexText,
-                expectOnly, undefined, /*inTypesVersionDirectory*/ true);
-        }
-
-        function getTsVersion(i: number): TsVersion {
-            return i === typesVersions.length
-                ? TypeScriptVersion.latest
-                : assertDefined(TypeScriptVersion.previous(typesVersions[i]));
+        // For example, typesVersions of [3.2, 3.5, 3.6] will have
+        // associated ts3.2, ts3.5, ts3.6 directories, for
+        // <=3.2, <=3.5, <=3.6 respectively; the root level is for 3.7 and above.
+        // so this code needs to generate ranges [lowest-3.2, 3.3-3.5, 3.6-3.6, 3.7-latest]
+        const lows = [TypeScriptVersion.lowest, ...typesVersions.map(next)]
+        const his = [...typesVersions, TypeScriptVersion.latest]
+        assert.strictEqual(lows.length, his.length)
+        for (let i = 0; i < lows.length; i++) {
+            const low = lows[i];
+            const hi = his[i];
+            const isLatest = hi === TypeScriptVersion.latest;
+            let versionPath = isLatest ? dirPath : joinPaths(dirPath, `ts${hi}`);
+            let versionIndexText = isLatest ? indexText : await readFile(joinPaths(versionPath, "index.d.ts"), "utf-8");
+            console.log('testing from', low, 'to', hi, 'in', versionPath)
```

Comment:
I'm not sure how noisy this will be in normal use of dtslint, but it surely makes the standalone output more readable.